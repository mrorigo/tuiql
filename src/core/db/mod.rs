/// Database Module
///
/// This module provides the core database functionality for TUIQL,
/// organized into focused submodules for better maintainability and
/// separation of concerns.
///
/// ## Architecture
///
/// The database layer is split into three main concerns:
/// - **Connection Management** (`connection.rs`): Handles database connections, state, and transactions
/// - **Schema Introspection** (`schema.rs`): Provides metadata about database structure
/// - **Query Execution** (`query.rs`): Handles SQL query execution and result formatting
///
/// ## Error Handling
///
/// All database operations use the standardized `TuiqlError` type for consistent error propagation.
///
/// ## Usage
///
/// Most common operations should use the higher-level functions from this module,
/// which automatically handle connection state and error conversion.
pub mod connection;
pub mod query;
pub mod schema;

pub use connection::*;
pub use query::*;
pub use schema::*;