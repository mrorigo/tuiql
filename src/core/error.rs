/// TUIQL Error Module
///
/// This module defines comprehensive error types for the TUIQL application.
/// It provides structured error handling with proper error propagation and
/// user-friendly error messages.
use thiserror::Error;

/// Comprehensive error type for the TUIQL application.
///
/// This enum covers all error scenarios that can occur within TUIQL:
/// - Database operations (connection, queries, transactions)
/// - Query execution and parsing
/// - Configuration management
/// - UI operations and export formats
/// - File system operations
/// - JSON parsing and validation
#[derive(Error, Debug)]
pub enum TuiqlError {
    /// Database-related errors from SQLite operations
    #[error("Database error: {0}")]
    Database(#[from] rusqlite::Error),

    /// SQL query errors (syntax, execution, missing tables, etc.)
    #[error("Query error: {0}")]
    Query(String),

    /// Configuration loading and validation errors
    #[error("Configuration error: {0}")]
    Config(String),

    /// UI-related errors (export formats, display issues)
    #[error("UI error: {0}")]
    Ui(String),

    /// File system and I/O errors
    #[error("I/O error: {0}")]
    Io(#[from] std::io::Error),

    /// JSON parsing and validation errors
    #[error("JSON error: {0}")]
    Json(#[from] serde_json::Error),

    /// Command validation and execution errors
    #[error("Command error: {0}")]
    Command(String),

    /// Transaction-related errors
    #[error("Transaction error: {0}")]
    Transaction(String),

    /// Schema-related errors
    #[error("Schema error: {0}")]
    Schema(String),

    /// Generic application errors for unexpected conditions
    #[error("Application error: {0}")]
    App(String),
}

/// Type alias for Result to use TuiqlError as the error type.
///
/// This provides a consistent error type across the entire application
/// instead of using `Result<T, String>` or mixed error types.
pub type Result<T> = std::result::Result<T, TuiqlError>;

/// Type alias for CLI command results that may need to return success/failure status
pub type CommandResult = Result<Option<String>>;

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_error_display() {
        let db_err = TuiqlError::Database(rusqlite::Error::ExecuteReturnedResults);
        assert!(db_err.to_string().contains("Database error"));

        let query_err = TuiqlError::Query("Syntax error".to_string());
        assert!(query_err.to_string().contains("Query error"));

        let config_err = TuiqlError::Config("Invalid config".to_string());
        assert!(config_err.to_string().contains("Configuration error"));
    }

    #[test]
    fn test_error_conversion() {
        // Test IO error conversion
        let io_err = std::io::Error::new(std::io::ErrorKind::NotFound, "file not found");
        let tuiql_err: TuiqlError = io_err.into();
        match tuiql_err {
            TuiqlError::Io(_) => {}
            _ => panic!("Expected IO error"),
        }

        // Test JSON error conversion
        let json_str = "{ invalid json }";
        let json_err: std::result::Result<serde_json::Value, serde_json::Error> = serde_json::from_str(json_str);
        let tuiql_err: TuiqlError = json_err.unwrap_err().into();
        match tuiql_err {
            TuiqlError::Json(_) => {}
            _ => panic!("Expected JSON error"),
        }
    }
}