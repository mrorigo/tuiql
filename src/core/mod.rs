/// Core Module for TUIQL
///
/// This module contains the fundamental components and utilities that form
/// the backbone of the TUIQL application. It provides shared infrastructure
/// for database operations, error handling, configuration management, and
/// other core functionality.

pub mod db;
pub mod error;

// Re-export commonly used types for convenience
pub use error::{TuiqlError, Result, CommandResult};