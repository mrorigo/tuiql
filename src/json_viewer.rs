/// JSON Tree Viewer Module for TUIQL
///
/// This module provides functionality to parse and display JSON data in a structured
/// tree format within the terminal. It supports nested objects and arrays, and provides
/// collapsible nodes for better navigation of complex JSON structures.
///
/// Key features:
/// - Parse JSON strings into tree structures
/// - Render tree with collapsible/expandable nodes
/// - Handle different JSON value types (objects, arrays, primitives)
/// - Integration with TUI for keyboard navigation
use crate::core::{Result, TuiqlError};

/// Represents different types of JSON values for display purposes.
#[derive(Debug, Clone, PartialEq)]
pub enum JsonValueType {
    Object,
    Array,
    String,
    Number,
    Boolean,
    Null,
}

/// A node in the JSON tree structure.
#[derive(Debug, Clone)]
pub struct JsonNode {
    pub key: String,             // The key (for objects) or index (for arrays)
    pub value: JsonValue,        // The actual value
    pub children: Vec<JsonNode>, // Child nodes (empty for primitives)
    pub expanded: bool,          // Whether this node is expanded (for collapsible nodes)
    pub depth: usize,            // How deep in the tree this node is
}

/// The different types of values a JSON node can hold.
#[derive(Debug, Clone)]
pub enum JsonValue {
    Object(serde_json::Map<String, serde_json::Value>),
    Array(Vec<serde_json::Value>),
    String(String),
    Number(serde_json::Number),
    Bool(bool),
    Null,
}

impl JsonValue {
    /// Get the type of this JSON value.
    pub fn value_type(&self) -> JsonValueType {
        match self {
            JsonValue::Object(_) => JsonValueType::Object,
            JsonValue::Array(_) => JsonValueType::Array,
            JsonValue::String(_) => JsonValueType::String,
            JsonValue::Number(_) => JsonValueType::Number,
            JsonValue::Bool(_) => JsonValueType::Boolean,
            JsonValue::Null => JsonValueType::Null,
        }
    }

    /// Get a display string for this value.
    pub fn display_value(&self) -> String {
        match self {
            JsonValue::Object(map) => format!("{{{} object}}", map.len()),
            JsonValue::Array(arr) => format!("[{} items]", arr.len()),
            JsonValue::String(s) => format!("\"{}\"", s),
            JsonValue::Number(n) => n.to_string(),
            JsonValue::Bool(b) => b.to_string(),
            JsonValue::Null => "null".to_string(),
        }
    }

    /// Check if this value has children (i.e., is collapsible).
    pub fn has_children(&self) -> bool {
        match self {
            JsonValue::Object(map) => !map.is_empty(),
            JsonValue::Array(arr) => !arr.is_empty(),
            _ => false,
        }
    }
}

/// The main JSON tree viewer component.
pub struct JsonTreeViewer {
    pub root: Option<JsonNode>,
    pub focused_path: Vec<String>, // Path to the currently focused node
}

impl Default for JsonTreeViewer {
    fn default() -> Self {
        Self::new()
    }
}

impl JsonTreeViewer {
    /// Create a new empty JSON tree viewer.
    pub fn new() -> Self {
        JsonTreeViewer {
            root: None,
            focused_path: Vec::new(),
        }
    }

    /// Load JSON data from a string.
    pub fn load_json(&mut self, json_str: &str) -> Result<()> {
        let parsed: serde_json::Value = serde_json::from_str(json_str).map_err(TuiqlError::Json)?;

        self.root = Some(self.build_tree_from_value("root".to_string(), parsed, 0));
        self.focused_path.clear();
        Ok(())
    }

    /// Build a tree structure from a serde_json Value.
    #[allow(clippy::only_used_in_recursion)]
    fn build_tree_from_value(
        &self,
        key: String,
        value: serde_json::Value,
        depth: usize,
    ) -> JsonNode {
        let (json_value, children) = match &value {
            serde_json::Value::Object(map) => {
                let obj_map = map.clone();
                let mut child_nodes = Vec::new();
                for (k, v) in &obj_map {
                    child_nodes.push(self.build_tree_from_value(k.clone(), v.clone(), depth + 1));
                }
                (JsonValue::Object(obj_map), child_nodes)
            }
            serde_json::Value::Array(arr) => {
                let arr_vec = arr.clone();
                let mut child_nodes = Vec::new();
                for (i, v) in arr_vec.iter().enumerate() {
                    child_nodes.push(self.build_tree_from_value(
                        i.to_string(),
                        v.clone(),
                        depth + 1,
                    ));
                }
                (JsonValue::Array(arr_vec), child_nodes)
            }
            serde_json::Value::String(s) => (JsonValue::String(s.clone()), Vec::new()),
            serde_json::Value::Number(n) => (JsonValue::Number(n.clone()), Vec::new()),
            serde_json::Value::Bool(b) => (JsonValue::Bool(*b), Vec::new()),
            serde_json::Value::Null => (JsonValue::Null, Vec::new()),
        };

        JsonNode {
            key,
            value: json_value,
            children,
            expanded: depth < 2, // Auto-expand first two levels
            depth,
        }
    }

    /// Render the JSON tree as a formatted string.
    pub fn render(&self) -> String {
        let mut output = String::new();
        if let Some(ref root) = self.root {
            self.render_node(&mut output, root, &Vec::new());
        } else {
            output.push_str("No JSON data loaded");
        }
        output
    }

    /// Render a single node and its children.
    fn render_node(&self, output: &mut String, node: &JsonNode, path: &[String]) {
        // Add indentation based on depth
        let indent = "  ".repeat(node.depth);
        let prefix = self.get_node_prefix(node);

        // Add focus indicator if this node is focused
        let focus_marker = if path == self.focused_path.as_slice() {
            "▶ "
        } else {
            "  "
        };

        // Format the node line
        output.push_str(&format!(
            "{}{}{} {}: {}\n",
            indent,
            focus_marker,
            prefix,
            node.key,
            node.value.display_value()
        ));

        // Render children if expanded
        if node.expanded && node.value.has_children() {
            for child in &node.children {
                let mut child_path = path.to_vec();
                child_path.push(child.key.clone());
                self.render_node(output, child, &child_path);
            }
        }
    }

    /// Get the appropriate prefix icon for a node type.
    fn get_node_prefix(&self, node: &JsonNode) -> &'static str {
        if node.value.has_children() {
            if node.expanded {
                "📂"
            } else {
                "📁"
            }
        } else {
            match node.value.value_type() {
                JsonValueType::String => "\"",
                JsonValueType::Number => "#",
                JsonValueType::Boolean => "✓",
                JsonValueType::Null => "∅",
                JsonValueType::Object | JsonValueType::Array => "📄", // Shouldn't happen for primitive types
            }
        }
    }

    /// Toggle the expanded state of a node at the given path.
    pub fn toggle_expanded(&mut self, path: &[String]) {
        if let Some(ref mut root) = self.root {
            toggle_node_expanded(root, path);
        }
    }
    /// Set the focused node path.
    pub fn set_focus(&mut self, path: Vec<String>) {
        self.focused_path = path;
    }

    /// Get a flat list of all expandable nodes for navigation.
    pub fn get_expandable_nodes(&self) -> Vec<Vec<String>> {
        let mut nodes = Vec::new();
        if let Some(ref root) = self.root {
            self.collect_expandable_nodes(root, &Vec::new(), &mut nodes);
        }
        nodes
    }

    /// Extract a sub-value from the JSON tree at the given path.
    pub fn extract_value(&self, path: &[String]) -> Option<&JsonValue> {
        if let Some(ref root) = self.root {
            return self.find_value_at_path(root, path);
        }
        None
    }

    /// Recursively collect all expandable nodes.
    #[allow(clippy::only_used_in_recursion)]
    fn collect_expandable_nodes(
        &self,
        node: &JsonNode,
        current_path: &[String],
        nodes: &mut Vec<Vec<String>>,
    ) {
        if node.value.has_children() {
            let mut path = current_path.to_vec();
            path.push(node.key.clone());
            nodes.push(path.clone());

            if node.expanded {
                for child in &node.children {
                    self.collect_expandable_nodes(child, &path, nodes);
                }
            }
        }
    }

    /// Recursively find a value at the given path.
    #[allow(clippy::only_used_in_recursion)]
    fn find_value_at_path<'a>(
        &self,
        node: &'a JsonNode,
        target_path: &[String],
    ) -> Option<&'a JsonValue> {
        if target_path.is_empty() {
            return Some(&node.value);
        }

        let current_key = &target_path[0];
        let remaining_path = &target_path[1..];

        for child in &node.children {
            if &child.key == current_key {
                return self.find_value_at_path(child, remaining_path);
            }
        }
        None
    }
}

/// Recursively find and toggle a node at the given path.
/// This is a standalone function to avoid borrowing issues within the impl.
fn toggle_node_expanded(node: &mut JsonNode, target_path: &[String]) {
    if target_path.is_empty() {
        if node.value.has_children() {
            node.expanded = !node.expanded;
        }
        return;
    }

    let current_key = &target_path[0];
    let remaining_path = &target_path[1..];

    for child in &mut node.children {
        if &child.key == current_key {
            toggle_node_expanded(child, remaining_path);
            return;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_snapshot;

    #[test]
    fn test_load_simple_json() {
        let mut viewer = JsonTreeViewer::new();
        let json = r#"{"name": "Alice", "age": 30, "active": true}"#;

        assert!(viewer.load_json(json).is_ok());
        assert!(viewer.root.is_some());

        let rendered = viewer.render();
        assert!(rendered.contains("name"));
        assert!(rendered.contains("Alice"));
        assert!(rendered.contains("age"));
        assert!(rendered.contains("30"));
        assert!(rendered.contains("active"));
        assert!(rendered.contains("true"));
    }

    #[test]
    fn test_load_nested_json() {
        let mut viewer = JsonTreeViewer::new();
        let json = r#"{
            "user": {
                "name": "Bob",
                "profile": {
                    "hobbies": ["reading", "coding"],
                    "address": {
                        "city": "NYC",
                        "zip": "10001"
                    }
                }
            },
            "tags": ["developer", "rust"]
        }"#;

        assert!(viewer.load_json(json).is_ok());
        let rendered = viewer.render();
        println!("Rendered output:\n{}", rendered); // Debug output
        assert!(rendered.contains("user"));
        // User object has name, profile - so {2 object}
        assert!(rendered.contains("{2 object}"));
        assert!(rendered.contains("tags"));
        assert!(rendered.contains("[2 items]"));
    }

    #[test]
    fn test_toggle_expanded() {
        let mut viewer = JsonTreeViewer::new();
        let json = r#"{"root": {"child": "value"}}"#;

        viewer.load_json(json).unwrap();
        let expandable = viewer.get_expandable_nodes();

        // Should have at least the root object
        assert!(!expandable.is_empty());

        // Find the path to root object and toggle it
        let root_path = expandable
            .into_iter()
            .find(|path| path.len() == 1 && path[0] == "root");
        if let Some(path) = root_path {
            viewer.toggle_expanded(&path);
            // Note: In a real TUI implementation, we would re-render to verify the change
        }
    }

    #[test]
    fn test_invalid_json() {
        let mut viewer = JsonTreeViewer::new();
        let invalid_json = r#"{"invalid": json}"#;

        assert!(viewer.load_json(invalid_json).is_err());
        assert!(viewer.root.is_none());
    }

    #[test]
    fn test_empty_viewer() {
        let viewer = JsonTreeViewer::new();
        let rendered = viewer.render();
        assert_eq!(rendered, "No JSON data loaded");
    }

    #[test]
    fn test_extract_value() {
        let mut viewer = JsonTreeViewer::new();
        let json = r#"{"users": [{"name": "Alice"}]}"#;

        viewer.load_json(json).unwrap();

        // Extract the users array
        let users_value = viewer.extract_value(&["users".to_string()]);
        assert!(users_value.is_some());

        if let Some(JsonValue::Array(arr)) = users_value {
            assert_eq!(arr.len(), 1);
        } else {
            panic!("Expected array");
        }
    }

    #[test]
    fn test_render_golden_simple_object() {
        let mut viewer = JsonTreeViewer::new();
        viewer
            .load_json(r#"{"name": "Alice", "age": 30, "active": true, "score": null}"#)
            .unwrap();
        assert_snapshot!("json_viewer_simple_object", viewer.render());
    }

    #[test]
    fn test_render_golden_nested_structure() {
        let mut viewer = JsonTreeViewer::new();
        viewer
            .load_json(
                r#"{
            "company": {
                "name": "TechCorp",
                "departments": [
                    {"name": "Engineering", "employees": 25},
                    {"name": "Sales", "employees": 15}
                ],
                "location": {"city": "San Francisco", "state": "CA"}
            },
            "tags": ["startup", "tech", "innovative"]
        }"#,
            )
            .unwrap();
        assert_snapshot!("json_viewer_nested_structure", viewer.render());
    }

    #[test]
    fn test_render_golden_array() {
        let mut viewer = JsonTreeViewer::new();
        viewer
            .load_json(
                r#"[
            {"id": 1, "product": "Laptop", "price": 999.99},
            {"id": 2, "product": "Mouse", "price": 29.99},
            {"id": 3, "product": "Keyboard", "price": 79.99}
        ]"#,
            )
            .unwrap();
        assert_snapshot!("json_viewer_array", viewer.render());
    }

    #[test]
    fn test_render_golden_empty_viewer() {
        let viewer = JsonTreeViewer::new();
        assert_snapshot!("json_viewer_empty", viewer.render());
    }

    #[test]
    fn test_render_golden_with_focus() {
        let mut viewer = JsonTreeViewer::new();
        viewer
            .load_json(r#"{"root": {"child1": "value1", "child2": {"grandchild": "value2"}}}"#)
            .unwrap();
        viewer.set_focus(vec![
            "root".to_string(),
            "child2".to_string(),
            "grandchild".to_string(),
        ]);
        assert_snapshot!("json_viewer_with_focus", viewer.render());
    }
}
