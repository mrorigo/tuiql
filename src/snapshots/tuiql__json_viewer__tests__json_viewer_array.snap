---
source: src/json_viewer.rs
expression: viewer.render()
---
â–¶ ðŸ“‚ root: [3 items]
    ðŸ“‚ 0: {3 object}
      # id: 1
      # price: 999.99
      " product: "Laptop"
    ðŸ“‚ 1: {3 object}
      # id: 2
      # price: 29.99
      " product: "Mouse"
    ðŸ“‚ 2: {3 object}
      # id: 3
      # price: 79.99
      " product: "Keyboard"
